### DISCUSSION TITLE
PipeEnd Printer

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-parsing#226] [Q&A] [answered]
Hi! Was wondering if PipeEnd can be made a printer in some (maybe all?) cases. 
I have the following parser that fails to compile with the following errors:
`Generic struct 'ParsePrint' requires that 'Parsers.PipeEnd<Substring.UTF8View>' conform to 'ParserPrinter'`
`Initializer 'init(with:)' requires that 'Parsers.PipeEnd<Substring.UTF8View>' conform to 'ParserPrinter'`

```swift
let testParser = ParsePrint {
    Prefix(1...) { $0 != .init(ascii: ">") }.pipe {
        PrefixUpTo("_".utf8)
        "_".utf8
        PrefixUpTo("_".utf8)
        "_".utf8
        PrefixUpTo("/".utf8)
        "/".utf8
    }
}
```

It can be fixed with the following workaround: 
```swift
let testParser = ParsePrint {
    Prefix(1...) { $0 != .init(ascii: ">") }.map(.substring).pipe {
        From(.utf8) {
            PrefixUpTo("_".utf8)
            "_".utf8
            PrefixUpTo("/".utf8)
            "/".utf8
        }
    }
}
```

However I don't immediately see why PipeEnd can't be a printer. A (perhaps naive) `PipeEnd` printer implementation fixes it:
```swift
extension Parsers.PipeEnd: ParserPrinter
{
    @inlinable
    public func print(_ output: (), into input: inout Input) throws { }
}
```
I might be missing a more subtle case where this fails however. Otherwise I'd be glad to make a PR.

**Comment by stephencelis** [comment, marked as answer]
@JaapWijnen I think this is an oversight (we've found we reach for `map` with conversions more often than `pipe`). If you'd like to PR, the implementation should probably be the exact same as `End`'s printer conformance.

**Reply by JaapWijnen** [reply]
You can't pass a parser as a conversion can you? (just confused by  "reaching for map more often than pipe" for a second)

**Reply by JaapWijnen** [reply]
The PR is up!
